### ISSUE TITLE
Some user-mode program call vmxxxx to generate #GP but captured by vmm

### ISSUE THREAD

**Issue by hzqst** [tandasat/HyperPlatform#26]
Recently I found some user-mode program running ical "vmcall" with an exception handler (SEH or VEH). if it's running under HyperPlatform's monitor, the "vmcall" will be handled by the vmexit handler, and the program won't generate a #GP so that it will not be working correctly.

**Comment by tandasat** [comment]
Good catch. Could you possible to provide to understand your expectation completely and implement fix correctly:

- the tested program or a test code snippet that you think HyperPlatform should handle proper
- brief test procedure if complicated
- brief description of expected result with the test procedure

I think what HyperPlatform should do is just delivering #GP just like when no VMM is installed, but I would like to understand the issue and impact clearer. Thank you for reporting the issue!

**Comment by hzqst** [comment]
```
#include "stdafx.h"
#include <windows.h>
#include <intrin.h>

int filter(EXCEPTION_POINTERS * pException, BOOL *pFound)
{
	auto code = *(BYTE*)pException->ContextRecord->Eip;

	printf("cpuid exception eip = %p code = %02x\n", pException->ContextRecord->Eip, code);

	//code != nop, in vmm
	if (code != 0x90)
		*pFound = TRUE;

	return EXCEPTION_EXECUTE_HANDLER;
}

int filter2(EXCEPTION_POINTERS * pException, BOOL *pFound)
{
	printf("vmcall exception eip = %p, exception = %X\n", pException->ContextRecord->Eip, pException->ExceptionRecord->ExceptionCode);
	
	//ExceptionCode == #GP, in real OS
	if (pException->ExceptionRecord->ExceptionCode == 0xC000001D)
		*pFound = FALSE;

	return EXCEPTION_EXECUTE_HANDLER;
}

BOOL vmcall_test()
{
	BOOL bFound = TRUE;
	__try
	{
		//vmcall
		__asm {
			_emit 0x0f;
			_emit 0x01;
			_emit 0xc1;
		}
	}
	__except (filter2(GetExceptionInformation(), &bFound))
	{
		;
	}
	return bFound;	
}

BOOL cpuid_test()
{
	BOOL bFound = FALSE;
	__try
	{
		__asm
		{
			mov eax, 0
			pushfd;
			or DWORD ptr[esp], 0x100;
			popfd;//set TF=1
			cpuid;
			nop;//normal TF
			int 3//in vmm...
		}
	}
	__except (filter(GetExceptionInformation(), &bFound))
	{
		;
	}
	return bFound;
}

int main()
{
	auto b = cpuid_test();
	if (b)
		printf("cpuid run in vmm\r\n");
	else
		printf("cpuid run in real\r\n");

	auto c = vmcall_test();
	if (c)
		printf("vmcall run in vmm\r\n");
	else
		printf("vmcall run in real\r\n");

	system("pause");
	return 0;
}
```

After tested in vmm and real OS,  I got different results:
![real](https://cloud.githubusercontent.com/assets/12287588/18034804/0749e898-6d79-11e6-8088-815ade926dfb.png)
![vmm](https://cloud.githubusercontent.com/assets/12287588/18034807/0bc45638-6d79-11e6-8e98-4f8e2de0639b.png)

I have no idea how to fix these "backdoors"... I don't want user-mode programs to know that they are in VMM.

**Comment by tandasat** [comment]
Thanks for the test code and test results, that is very helpful. 

I confirmed behavior as well. I will investigate a cause and consider possible fix so that HyperPlatform does not introduce the difference.
![issue_26](https://cloud.githubusercontent.com/assets/1620923/18169346/946be8f6-700e-11e6-9dff-e0f2648bb550.png)

**Comment by f74108** [comment]
enable rdtsc wbinvd ，Win7 X86 PAE 30 minutes will lead to blue screen

**Comment by tandasat** [comment]
@f74108 do you think that is the same issue as one discussed in this thread? If not, could you please file a separate issue for this?

**Comment by tandasat** [comment]
Hi @hzqst  can you try out the latest code including a fix for this issue (6e3e139) and let me know if that works as expected?

**Comment by hzqst** [comment]
@tandasat 
I tried the lastest code in win7 x64 7600 with vmware 12.0.0 and got this
![why](https://cloud.githubusercontent.com/assets/12287588/18223322/86998376-71e6-11e6-8b4c-4588c5d6d05e.png)

vmcall_test is ok now, but cpuid_test still return an unexpected result.

**Comment by tandasat** [comment]
Thanks for testing. My bad, you were reporting the failure in the CPUID test from the beginning.

I was not able to reproduce the issue with my VMware 11, but could with my laptop interestingly. 
![issue_26](https://cloud.githubusercontent.com/assets/1620923/18225519/87b75bc2-71a8-11e6-9d0b-3cf5677472f0.png)

Did you have this line in your VM's VMX file? 

    hypervisor.cpuid.v0 = "FALSE"

I do have it, but it you do not, I will take it  off and test on the VM again. I am hoping to be able to reproduce the issue on my VMware if possible.

**Comment by hzqst** [comment]
I do not have
`hypervisor.cpuid.v0 = "FALSE"`
and when I add this to vmx file it still produce the issue in cpuid_test();

**Comment by tandasat** [comment]
Thank you for checking. I will test with a real device than.

**Comment by tandasat** [comment]
do you know any resources on how and why the CPUID based detection works?

**Comment by hzqst** [comment]
```
__asm
        {
            mov eax, 0
            pushfd;
            or DWORD ptr[esp], 0x100;
            popfd;//set TF=1
            cpuid;
            nop;//normal TF
            int 3//in vmm...
        }
```

what cpu do in real os:
execute CPUID->**set EIP at nop**->raise an single step exception(#DB) with error code 0x80000004->set TF=0

what cpu do in vmm:
execute cpuid->cpuid is handled by vm-exit handler->you advance the EIP by one instrution in the handler (so EIP would be at nop)->execute nop->**set EIP at "int 3"**->raise an single step exception->set TF=0

that's difference between vmm and real os.

**Comment by tandasat** [comment]
Thanks for detailed explanation. I am looking for info on which open source VMM exhibit this difference and how they handle it if not. from source code, I coudld not tell how KVM and VirtualBox avoid this difference yet.

**Comment by asamy** [comment]
@tandasat  the solution is to inject #DB when Guest's TF is set, and adjusting RIP normally, also writing the instruction length of CPUID in the vm entry instruction len which is 2.

**Comment by tandasat** [comment]
@asamy Thank you a lot for the sample code! I implemented it with 3518e481f5422b7d7202e3559347304d934db87b. 

@hzqst Can you please test the latest code and let me know if it looks good?

**Comment by hzqst** [comment]
tested, it works well.
